google-cloud-spanner/src/test/java/com/google/cloud/spanner/LongRunningSessionsBenchmark.java

```diff
@@ -0,0 +1,317 @@
+    List<ListenableFuture<?>> futures = new ArrayList<>(totalWrites);
+    for (int i = 0; i < totalWrites; i++) {
+      futures.add(
+          service.submit(
+              () -> {
+                // introduce random sleep times so that some sessions are long-running sessions
+                randomWaitForMockServer(server);
+                client.executePartitionedUpdate(StandardBenchmarkMockServer.UPDATE_STATEMENT);
+              }));
+    }
+    Futures.allAsList(futures).get();
+    service.shutdown();
+    assertThat(pool.numLeakedSessionsRemoved())
+        .isEqualTo(0); // no sessions should be cleaned up in case of partitioned updates.
+  }
+
+  /**
+   * Measures the time needed to execute a burst of read and write requests.
+   *
+   * <p>Some read requests will be long-running and will cause session leaks. Such sessions will be
+   * removed by the session maintenance background task if SessionPool Option
+   * ActionOnInactiveTransaction is set as WARN_AND_CLOSE.
+   *
+   * <p>Some write requests will be long-running. The test asserts that no sessions are removed by
+   * the session maintenance background task with SessionPool Option ActionOnInactiveTransaction set
+   * as WARN_AND_CLOSE. This is because PDML writes are expected to be long-running.
+   *
+   * @param server
+   * @throws Exception
+   */
+  @Benchmark
+  public void burstReadAndWrite(final BenchmarkState server) throws Exception {
+    int totalWrites = server.maxSessions * 4;
+    int totalReads = server.maxSessions * 4;
+    int parallelThreads = server.maxSessions * 2;
+    final DatabaseClient client =
+        server.spanner.getDatabaseClient(DatabaseId.of(TEST_PROJECT, TEST_INSTANCE, TEST_DATABASE));
+    SessionPool pool = ((DatabaseClientImpl) client).pool;
+    assertThat(pool.totalSessions()).isEqualTo(server.minSessions);
+
+    ListeningScheduledExecutorService service =
+        MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(parallelThreads));
+    List<ListenableFuture<?>> futures = new ArrayList<>(totalReads + totalWrites);
+    for (int i = 0; i < totalWrites; i++) {
+      futures.add(
+          service.submit(
+              () -> {
+                // introduce random sleep times so that some sessions are long-running sessions
+                randomWaitForMockServer(server);
+                client.executePartitionedUpdate(StandardBenchmarkMockServer.UPDATE_STATEMENT);
+              }));
+    }
+    for (int i = 0; i < totalReads; i++) {
+      futures.add(
+          service.submit(
+              () -> {
+                Thread.sleep(RND.nextInt(RND_WAIT_TIME_BETWEEN_REQUESTS));
+                try (ResultSet rs =
+                    client.singleUse().executeQuery(StandardBenchmarkMockServer.SELECT1)) {
+                  while (rs.next()) {
+                    // introduce random sleep times to have long-running sessions
+                    randomWait();
```

Comment:
Pure theoretically this random function could end up never adding a long wait time if `random.nextBool` returns false every time. Would it maybe make sense to let this method update an atomic integer every time it sets  a long wait time, and then at the end of the test verify that the number of sessions that was removed is equal to that atomic integer (instead of just checking that it's greater than zero, which could in theory fail)?

Reply:
That's a very good idea. Let me do this. It will make the benchmark even more precise.